cluster_pack/filesystem.py

```diff
@@ -203,6 +230,19 @@ def put(self, filename: str, path: str, chunk: int = 2**16) -> None:
                     target.write(out)
         _preserve_acls(self.base_fs, filename, path)
 
+    def put_atomic(self,
+         source_path: str,
+         destination_path: str) -> None:
+        tmp_upload_dir = f"{destination_path}.{uuid.uuid4()}.tmp"
+        self.put(source_path, tmp_upload_dir)
+        try:
+            self.mv(tmp_upload_dir, destination_path)
+        except OSError as err:
+            raise FileExistsError(f"File exist at destination {destination_path}", err)

```

Comment:
My understanding is that the existing put just overrides the file. It means `put` and `atomic_put` won't behave the same way.
My suggestion would be to delete before if the file exists, or we load the file and check its content before deleting.
In any case we can't do an atomic operation with HDFS, mv is not atomic either but our assumption is that delete and mv are significantly faster than the upload with put.